Distance tests to z=1050 for massive neutrinos and curvature

tilmantroester opened the pull request:
This PR aims to address issue #176. It includes tests at 10 logarithmically spaced redshifts between 0.01 and 1050 for:

- no curvature, no massive neutrinos, 3 massless neutrinos
- positive curvature, no massive neutrinos, 3 massless neutrinos
- negative curvature, no massive neutrinos, 3 massless neutrinos
- no curvature, 1 massive neutrinos, 2 massless neutrinos
- no curvature, 3 massive neutrinos (2 different masses), no massless neutrinos
- no curvature, 3 massive neutrinos (3 different masses), no massless neutrinos
- no curvature, no massive neutrinos, 6 massless neutrinos
- negative curvature, 2 massive neutrinos, 4 massless neutrinos
- positive curvature, 3 massive neutrinos, 3 massless neutrinos

The distances are compared against a benchmark created with CLASS. The notebook to create the benchmark is included in the PR.

In order for the tests to pass, the calculation of Omega_nu_rel needs to be fixed (see issue #372), which currently breaks some tests (CMB lensing and 3D correlation function), however.

The new tests pass with tolerance of 1e-5 if the first redshift (z=0.01) is omitted, such that the tests are from z=0.03 to z=1050. For z=0.01, the tolerance needs to be relaxed to 1e-4.

slosar commented:
This is great! So, there was a bug in h in the neutrino branch that you now fixed?
One question: this z=0.01 doesn't matter, but it is annoying. It is also possible it is a class issue, rather than our issue. Can you see if it can get cured using either improving the number of points in our spline or class spline? Should be minimal change. Otherwise I'm happy with this to go forward. Great work!!

tilmantroester commented:
I changed the `back_integration_stepsize` parameter in CLASS to something smaller and now get 1e-6 agreement with CCL everywhere (which is also the precision setting for distance integrals in CCL, I think). The test still uses 1e-5, in case we add more challenging models later on.

c-d-leonard commented on src/ccl_core.c:
```diff
@@ -243,16 +243,22 @@ void ccl_parameters_fill_initial(ccl_parameters * params, int *status)
-  //Get the relativistic neutrino Omega_nu. It's more efficient to get this in-line, to avoid computing the phase-space integral if not necessary.
-  double Tnu= (params->T_CMB) *pow(4./11.,1./3.); 
-  params-> Omega_n_rel = params->N_nu_rel* 8. * pow(M_PI,5) *pow((KBOLTZ/ HPLANCK),3)* KBOLTZ/(15. *pow( CLIGHT,3))* (8. * M_PI * GNEWT) / (3. * 100.*100.*1000.*1000. /MPC_TO_METER /MPC_TO_METER  * CLIGHT * CLIGHT)  * Tnu * Tnu * Tnu * Tnu *7./8.;
+  params->N_nu_rel = params->Neff - params->N_nu_mass * pow(TNCDM, 4) / pow(4./11.,4./3.);
```
I'm okay with chaning TNCDM * TNCDM * TNCDM * TNCDM to pow(TNCDM,4) if its more efficient - have you tested this? If its for readability at the expense of efficiency I would prefer to keep it as in, but others may disagree.

tilmantroester replied:
I'm of the opinion that readability should always come before efficiency unless code has been shown to be a mission-critical bottleneck.
For this specific example, I'd be very surprised if the compiler doesn't optimise both expressions to the same machine code anyway. Since this is function is just called once per cosmology, efficiency shouldn't be an issue either.

c-d-leonard commented on src/ccl_core.c:
```diff
@@ -243,16 +243,22 @@ void ccl_parameters_fill_initial(ccl_parameters * params, int *status)
+  double rho_nu_rel = params->N_nu_rel* 7.0/8.0 * 4. * STBOLTZ / pow(CLIGHT, 3) * pow(T_nu, 4);
+  params-> Omega_n_rel = rho_nu_rel/rho_crit;
+    
   // If non-relativistic neutrinos are present, calculate the phase_space integral.
```
I checked the math on this new expression vs the old one and I agree it the new Omega_n_rel expression solves the issues of the missing h^2 factor.

tilmantroester commented:
I've now fixed the tests and moved the new tests to ccl_test_distances.py. Only the CMB lensing test was affected, the 3d correlation test failing was due to an interaction between the two tests. This has been addressed now but the way CTEST deals with clean-up should probably be revisited at some point.

c-d-leonard commented on tests/ccl_test_distances.py:
```diff
@@ -257,7 +308,24 @@ def compare_distances_mnu_hiz(z, chi_bench,dm_bench, Omega_v, w0, wa, Neff_mnu,
 
     assert_allclose(dm, dm_bench[a_not_one], atol=1e-3, rtol = DISTANCES_TOLERANCE_MNU)
+
+@decorators.slow 
```
We probably don't need to label this as slow - usually calling distances is pretty fast? We only want to use this label for things that take a really long time to compute and significantly slow down running the test script.
